website/blog/2022-10-31-simple-1.mdx

```diff
@@ -0,0 +1,102 @@
+
+<figure>
+<img src="/img/simple1_1.png"/>
+</figure>
+
+The user only needs to implement the PlusFunction struct shown in the graph above, which is then expanded using template expansion to a vector function.
+
+Writing simple functions is quite simple and easy; however it's not as powerful as writing vector functions.
+For example the function _map_keys_ can be implemented in O(1) as a vector function by moving the keys vector; this is not possible to express as a simple function.
+
+Another limitation is that when using the simple interface, authors do not have access to the encodings of the input vectors, nor control over the encoding of the result vector. Hence, do not have the power to optimize the code for specific input encodings or optimize it by generating specific output encodings. The _array_sort_ function for instance does not need to re-order the elements and copy them during sorting; instead it can generate a dictionary vector as an output, which is something not expressible as a simple function.
+
+On the other hand, writing vector functions has some drawbacks:
+
+- **Complexity** : Requires an understanding of Velox vectorized data representation and encodings, which requires additional work for our customers, specially those without DB background. And writing optimized vector functions requires even deeper understanding.
+- **Repetition** : Involves repeated efforts and code; in each function vectors are read in the same way, same set of optimizations are applied ... etc.
+- **Reliability** : More code means more bugs, especially in such a complex context.
+
+In the ideal world we would like to add most of the optimization that someone can do in a vector function to the simple functions adapter, so it would be enabled automatically. We have identified a number of optimizations that apply to all functions and implemented these generically in the SimpleFunctionAdapter. In this way, we can achieve the good of the two worlds and gain **Simplicity**, **Efficiency** and **Reliability** for most functions.
+
+In the past year we have been working on several improvements on the simple function interface on both the `expressivity` and `performance` axes that we will discuss in this series of notes. In this blogpost, we will talk about
+some of the general optimizations that we have in the adapter.
+
+## General optimizations
+
+#### Vector reuse
+If the output type matches one of the input types, and the input vector is to die after the function invocation, then it is possible to reuse it for the results instead of allocating a new vector.
+
+#### Bulk null setting
+Nulls are represented in a bit vector, hence, writing each bit can be expensive specially for primitive operations (like plus and minus). One optimization is
+to optimize for the `not null` case, and bulk setting the nulls to not null (64 bits in one instruction). After that during the computation, only if the results are null, the null bit is set to null.
```

Comment:
don't we also do this for other cases as well? I mean, bulk setting null values based on the current active rows?